share/shwap/p2p/bitswap/block_store.go

```diff
@@ -0,0 +1,113 @@
+package bitswap
+
+import (
+	"context"
+	"fmt"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+
+	eds "github.com/celestiaorg/celestia-node/share/new_eds"
+	bitswappb "github.com/celestiaorg/celestia-node/share/shwap/p2p/bitswap/pb"
+)
+
+// Accessors abstracts storage system that indexes and manages multiple eds.Accessors by network height.
+type Accessors interface {
+	// Get returns an Accessor by its height.
+	Get(ctx context.Context, height uint64) (eds.Accessor, error)
+}
+
+// Blockstore implements generalized Bitswap compatible storage over Shwap containers
+// that operates with Block and accesses data through Accessors.
+type Blockstore struct {
+	Accessors
+}
+
+func (b *Blockstore) getBlock(ctx context.Context, cid cid.Cid) (blocks.Block, error) {
+	spec, ok := specRegistry[cid.Prefix().MhType]
+	if !ok {
+		return nil, fmt.Errorf("unsupported Block type: %v", cid.Prefix().MhType)
+	}
+
+	blk, err := spec.builder(cid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build a Block for %s: %w", spec.String(), err)
+	}
+
+	eds, err := b.Accessors.Get(ctx, blk.Height())
+	if err != nil {
+		return nil, fmt.Errorf("getting EDS Accessor for height %v: %w", blk.Height(), err)
+	}
+
+	if err = blk.Populate(ctx, eds); err != nil {
+		return nil, fmt.Errorf("failed to populate Shwap Block on height %v for %s: %w", blk.Height(), spec.String(), err)
+	}
+
+	containerData, err := blk.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("marshaling Shwap container: %w", err)
+	}
+
+	blkProto := bitswappb.Block{
+		Cid:       cid.Bytes(),
+		Container: containerData,
+	}
+
+	blkData, err := blkProto.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("marshaling Bitswap Block protobuf: %w", err)
+	}
+
+	bitswapBlk, err := blocks.NewBlockWithCid(blkData, cid)
+	if err != nil {
+		return nil, fmt.Errorf("assembling Bitswap block: %w", err)
+	}
+
+	return bitswapBlk, nil
+}
+
+func (b *Blockstore) Get(ctx context.Context, cid cid.Cid) (blocks.Block, error) {
+	blk, err := b.getBlock(ctx, cid)
+	if err != nil {
+		log.Errorf("blockstore: getting local block(%s): %s", cid, err)
+		return nil, err
+	}
+
+	return blk, nil
+}
+
+func (b *Blockstore) GetSize(ctx context.Context, cid cid.Cid) (int, error) {
+	// TODO(@Wondertan): There must be a way to derive size without reading, proving, serializing and
+	//  allocating Sample's block.Block or we could do hashing
+	// NOTE:Bitswap uses GetSize also to determine if we have content stored or not
+	// so simply returning constant size is not an option
+	blk, err := b.Get(ctx, cid)
+	if err != nil {
+		return 0, err
+	}
+	return len(blk.RawData()), nil
+}
+
+func (b *Blockstore) Has(ctx context.Context, cid cid.Cid) (bool, error) {
+	_, err := b.Get(ctx, cid)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
```

Comment:
This also can be optimised to work faster than full Get

Reply:
We could, but its not a hot path and optimizing GetSize is more important

Reply:
It is quite simple, just need to call Has on AccessorGetter. There will be optimised Has on store

Reply:
Has is not used by Bitswap(just checked), so we can avoid this or maybe even remove Has implementation